frontend/src/components/UserUpdateRoles.vue

```diff
@@ -1,210 +1,384 @@
+        { text: "Role", value: "roleArray" },
+        { text: "Last Log In", value: "lastLogin"},
+        { text: "Actions", value: "actions", sortable: false },
+      ],
+      clientWithoutRoles: [],
       selectedClient: null,
       clientRoles: [],
       effectiveClientRoles: [],
-      selectedRoles: []
+      selectedRoles: [],
+      ClientsWithEffectiveRoles: [],
+      rolesLoaded: false,
+      isEdit: false,
     };
   },
   async created() {
-    //TODO error handling
-    await this.getClients();
+   this.loadUserRoles();
   },
   computed: {
     itemsInColumn() {
-      return Math.ceil(this.clientRoles.length / this.numberOfClientRoleColumns);
+      return Math.ceil(
+        this.clientRoles.length / this.numberOfClientRoleColumns
+      );
     },
     numberOfClientRoleColumns() {
-      return (this.clientRoles.length > 10) ? 2 : 1
+      return this.clientRoles.length > 10 ? 2 : 1;
     },
-    hasRoleForManageUserRoles: function() {
+    hasRoleForManageUserRoles: function () {
       const umsClientId = "USER-MANAGEMENT-SERVICE";
       const manageUserRolesName = "manage-user-roles";
 
-      return !!this.$keycloak.tokenParsed.resource_access[umsClientId].roles.includes(manageUserRolesName)
-    }
+      return !!this.$keycloak.tokenParsed.resource_access[
+        umsClientId
+      ].roles.includes(manageUserRolesName);
+    },
   },
   methods: {
-    roleArrayPosition: function(col, item) {
-      return (col - 1) * (this.itemsInColumn) + item - 1;
+    addRoles() {
+      this.selectedClient = null;
+      this.isEdit = false;
+      this.dialog = true;
+    },
+    editRoles: function (client) {
+      this.selectedClient = client;
+      this.isEdit = true;
+      this.dialog = true;
+      this.getUserClientRoles();
     },
-    getClients: function() {
-      return ClientsRepository.get()
-        .then(response => {
-          this.clients = response.data;
+    close() {
+      this.dialog = false;
+    },
+    roleArrayPosition: function (col, item) {
+      return (col - 1) * this.itemsInColumn + item - 1;
+    },
+    addClientRoles: function () {
+      this.isEdit = false;
+      this.getUserClientRoles();
+    },
+    loadUserRoles: function () {
+      this.rolesLoaded = false;
+
+      let results = [];
+      let lastLoginMap = [];
+      let clientsNoRolesAssigned = [];
+
+      UsersRepository.getUserLogins(this.userId).then((lastLogins) => {
+        lastLoginMap = lastLogins.data;
+      });
+
+      ClientsRepository.get()
+        .then((allClients) => {
+          Promise.all(
+            allClients.data.map((client) => {
+              return UsersRepository.getUserEffectiveClientRoles(
+                this.userId,
+                client.id
+              ).then((clientRoles) => {
+                clientRoles["clientRepresentation"] = client;
+                return clientRoles;
+              });
+            })
+          )
+            .then(function (rolesArray) {
```

Comment:
I know this is from the copy/paste of the original method, and not something you added, but lets update this to use arrow notation like the rest of our methods
e.g. `(rolesArray) => {`